atomic_reactor/plugins/pre_resolve_module_compose.py

```diff
@@ -0,0 +1,272 @@
+    'args': {'module_name': 'myapp',
+             'module_stream': 'f26',
+             'module_version': '20170629185228',
+             'odcs_url': 'https://odcs.fedoraproject.org/odcs/1'},
+             'pdc_url': 'https://pdc.fedoraproject.org/rest_api/v1',}
+}
+"""
+
+import os
+import re
+from modulemd import ModuleMetadata
+from pdc_client import PDCClient
+import requests
+import time
+
+from atomic_reactor.plugin import PreBuildPlugin
+
+
+class ModuleInfo(object):
+    def __init__(self, name, stream, version, mmd):
+        self.name = name
+        self.stream = stream
+        self.version = version
+        self.mmd = mmd
+
+
+class ComposeInfo(object):
+    def __init__(self, compose_id, base_module, modules, repo_url):
+        self.compose_id = compose_id
+        self.base_module = base_module
+        self.modules = modules
+        self.repo_url = repo_url
+
+
+WORKSPACE_SOURCE_KEY = 'compose_info'
+
+
+def get_compose_info(workflow):
+    key = ResolveModuleComposePlugin.key
+    if key not in workflow.plugin_workspace:
+        return None
+    return workflow.plugin_workspace[key].get(WORKSPACE_SOURCE_KEY, None)
+
+
+def set_compose_info(workflow, source):
+    key = ResolveModuleComposePlugin.key
+
+    workflow.plugin_workspace.setdefault(key, {})
+    workspace = workflow.plugin_workspace[key]
+    workspace[WORKSPACE_SOURCE_KEY] = source
+
+
+class ODCSClient(object):
+    def __init__(self, url, insecure=False, token=None, log=None):
+        if url.endswith('/'):
+            self.url = url
+        else:
+            self.url = url + '/'
+        self.insecure = insecure
+        self.token = token
+        self.log = log
+
+    def _auth_headers(self):
+        headers = {}
+        if self.token:
+            headers['OIDC_access_token'] = self.token
+            headers['OIDC_CLAIM_scope'] = ('openid https://id.fedoraproject.org/scope/groups '
+                                           'https://pagure.io/odcs/new-compose '
+                                           'https://pagure.io/odcs/renew-compose '
+                                           'https://pagure.io/odcs/delete-compose')
+            # FIXME: passing the claim scope in the request doesn't make any sense,
+            #   because we're not to be trusted for what scopes we've claimed, but
+            #   match what the server wants.
+
+        return headers
+
+    def start_compose(self, source_type, source):
+        body = {
+            'source': {
+                'type': source_type,
+                'source': source
+            }
+        }
+
+        self.log.info("Starting compose for source_type={source_type}, source={source}"
+                      .format(source_type=source_type, source=source))
+        response = requests.post(self.url + 'composes/',
+                                 json=body,
+                                 headers=self._auth_headers())
+        response.raise_for_status()
```

Comment:
Are there any circumstances in which this POST request can be safely retried?

Reply:
Define "safe" :-) It's clearly safe to retry on a connection error. On other errors, it's hard to say - nothing horrible will go wrong if multiple composes are queued, but extra work will be done. 

(To be precise, only composes in the *done* state are reused, so additional copies of the compose are started until the first copy completes - but there's a configurable limit to how many composes are run in parallel - by default 2. If we retry at 5 seconds, 10 seconds, and 20 seconds, as I believe the retry default is, probably at most one extra compose will be done.)